examples/gp.py

```diff
@@ -123,6 +125,7 @@ def main(args):
     parser = argparse.ArgumentParser(description="Gaussian Process example")
     parser.add_argument("-n", "--num-samples", nargs="?", default=1000, type=int)
     parser.add_argument("--num-warmup", nargs='?', default=1000, type=int)
+    parser.add_argument("--num-chains", nargs='?', default=1, type=int)
```

Comment:
+1. I'm kind of surprised that all of these examples run with multiple chains with relatively little change to the codebase. How does the runtime compare now for 1 vs 4 chains for instance?

Reply:
FYI, I compare 1 (with the fastest variant progbar=None) vs 4 chains vs 12 chains

+ BNN example: 52s vs 56s vs 58s

+ GP example: 17s vs 20s vs 29s

To be fair, I should test with the slowest single chain and remove the `print_summary` process. But I hope these numbers are enough to get an overall idea.

Reply:
In GP example, maybe intra-op parallelism play an important role. :)

Reply:
The numbers look great to me. By forcing the device count to 2, are we disabling intra-op parallelism? If that's the case, maybe we are getting good results on these examples due to small data size, and something like covtype will be much slower, in comparison.

Reply:
Yeah, let me test that example too.

Reply:
+ Covtype example (HMC, num_steps=10, num_samples=100): 30s vs 34s vs 44s.

Somehow, the pattern is similar to GP. They are 1.5-2x slower in 12 cores, but consider the benefit that we get 12x more samples, that is a good trade-off.

Testing in covtype example, I'm observing that we get different summary results for 1 chain vs 4 chains. That may be just a random seed issue or a fast-math-mode issue. I'll investigate why but it does not related to this PR.

Reply:
That looks great, actually! I am quite surprised that there's that much headroom by including 12 cores since I would have thought that on covtype all the cpu cores would be utilized already. Will play around with it to see what's happening later.